File: Python/ch6/foreach_sample.py
import collections
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.streaming.context import StreamingContext

Person = collections.namedtuple('Person', 'name age')

conf = SparkConf()
sc = SparkContext(master="local[*]", appName="ForeachSample", conf=conf)
ssc = StreamingContext(sc, 1)

rdd1 = ssc.sparkContext.parallelize([Person("P1", 20)])
rdd2 = ssc.sparkContext.parallelize([Person("P2", 10)])
queue = [rdd1, rdd2]

ds = ssc.queueStream(queue)


def dataFrameOp(rdd):
    spark = SparkSession \
        .builder \
        .appName("sample") \
        .master("local[*]") \
        .getOrCreate()
    if rdd.count() > 0:
        df = spark.createDataFrame(rdd)
        df.select("name", "age").show()

ds.foreachRDD(lambda rdd: dataFrameOp(rdd))

ssc.start()
ssc.awaitTermination()

File: src/main/java/com/wikibooks/spark/ch6/ForeachSample.java
package com.wikibooks.spark.ch6;

import org.apache.commons.lang3.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.Optional;
import org.apache.spark.api.java.function.*;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import scala.Tuple2;

import java.util.*;

public class ForeachSample {

  public static void main(String[] args) throws Exception {

    SparkConf conf = new SparkConf().setMaster("local[*]").setAppName("ForeachSample");
    JavaStreamingContext ssc = new JavaStreamingContext(conf, Durations.seconds(1));
    JavaSparkContext sc = ssc.sparkContext();

    JavaRDD<Person> rdd1 = sc.parallelize(Arrays.asList(new Person("P1", 20)));
    JavaRDD<Person> rdd2 = sc.parallelize(Arrays.asList(new Person("P2", 10)));

    Queue<JavaRDD<Person>> queue = new LinkedList<>(Arrays.asList(rdd1, rdd2));
    JavaDStream<Person> ds = ssc.queueStream(queue);

    ds.foreachRDD(new VoidFunction<JavaRDD<Person>>() {
      @Override
      public void call(JavaRDD<Person> rdd) throws Exception {
        SparkSession spark = SparkSession
                .builder()
                .appName("Sample")
                .master("local[*]")
                .getOrCreate();
        Dataset<Row> df = spark.createDataFrame(rdd, Person.class);
        df.select("name", "age").show();
      }
    });

    ssc.start();
    ssc.awaitTermination();
  }
}